### DISCUSSION TITLE
Reading an object with colon in its name

### DISCUSSION THREAD

**Discussion by plasorak** [scikit-hep/uproot5#365] [Q&A]
Hi all,

I'm trying to read an object which has colon in its name. So doing something like:
`data = up.open("path/to/file.root")["that:histo:with:colon:in:the:middle"]`
That exits with the following error:
```  File "/usr/local/lib/python3.9/site-packages/uproot/reading.py", line 1962, in __getitem__
    step = step[item]
  File "/usr/local/lib/python3.9/site-packages/uproot/reading.py", line 1962, in __getitem__
    step = step[item]
  File "/usr/local/lib/python3.9/site-packages/uproot/reading.py", line 1962, in __getitem__
    step = step[item]
  [Previous line repeated 992 more times]
  File "/usr/local/lib/python3.9/site-packages/uproot/reading.py", line 1945, in __getitem__
    if ":" in item and item not in step:
  File "/usr/local/lib/python3.9/site-packages/uproot/reading.py", line 1810, in __contains__
    self.key(where)
  File "/usr/local/lib/python3.9/site-packages/uproot/reading.py", line 1887, in key
    where = uproot._util.ensure_str(where)
  File "/usr/local/lib/python3.9/site-packages/uproot/_util.py", line 72, in ensure_str
    if not py2 and isinstance(x, bytes):
RecursionError: maximum recursion depth exceeded while calling a Python object
```
How would I do that?
Thanks.

**Comment by jpivarski** [comment]
Short answer: put the file name in a `pathlib.Path`. These are not colon-split. In this mode, you can't put the object path in the same string, it has to be `uproot.open(pathlib.Path(filename))[object_path_here]`.

For more details, the thread where we discussed the design of this is #79.

**Reply by plasorak** [reply]
That works for the file name, but not for the _object_ path right?

**Reply by jpivarski** [reply]
Oh, I missed that. I hadn't realized that colons would be legal characters in objects—ah, it's TBranches that use colon as part of its constructor syntax.

Fortunately (really lucky, actually, the `TDirectory.key` method ([here](https://github.com/scikit-hep/uproot4/blob/e57e9ac2e06e4dc0d4e777191825fe014549401f/src/uproot/reading.py#L1926-L1988)) doesn't parse colons, so you can do

```python
data = uproot.open("path/to/file.root").key("that:histo:with:colon:in:the:middle").get()
```
